awsebcli: init at 3.10.5

eqyiel opened the pull request:
###### Motivation for this change

I wanted to use `awsebcli` at work and it wasn't in `nixpkgs` yet.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Also:

- moved `colorama_3_7` from the awscli derivation into `python-packages` since both `awscli` and `awsebcli` depend on it.
- refactored some of the parent derivations that I needed to override in order to use `fetchPypi`, which AFAICT is preferred for python packages (https://github.com/NixOS/nixpkgs/pull/25202#discussion_r113117919, https://github.com/NixOS/nixpkgs/pull/26078).
- I tried to find what is the current best practice regarding versioning derivation names but it appears that although there is [a section in the manual](https://nixos.org/nixpkgs/manual/#sec-package-naming) about it there is a lot of bikeshedding still (https://github.com/NixOS/nixpkgs/issues/17625), so I just went with what I think is sensible based on the names of other derivations in `pythonPackages` (`package_V_V`).

eqyiel commented:
Rebased!

FRidh commented on pkgs/tools/virtualization/awsebcli/default.nix:
```diff
@@ -0,0 +1,94 @@
+{ stdenv, fetchurl, pythonPackages }:
+let
+  botocore = pythonPackages.botocore.overrideAttrs (oldAttrs: rec {
```
Furthermore, due to the amount of overrides and the possible need of doing deep overrides, i suggest overriding the package set. E.g.

```
let

  mypython = python.override { 
    packageOverrides = self: super: {
      botocore = super.botocore.overridePythonAttrs (oldAttrs: {
        ...
      });  
      docker = super.docker.overridePythonAttrs( oldAttrs: {
        ...
      });
    };
  };
in ...
```

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -3093,6 +3086,17 @@ in {
   colorama = callPackage ../development/python-modules/colorama { };
 
+  # Needed by awscli and awsebcli.
+  colorama_3_7 = self.colorama.overrideAttrs (oldAttrs: rec {
```
overridePythonAttrs

FRidh reviewed (changes requested):
`package_V_V` is indeed the way forward.

eqyiel commented:
@FRidh thank you for the review, I've tried to address those points and updated the PR.

Overriding the package set is a lot nicer!

I also rebased onto master again.

I plan to add myself as a maintainer for the new packages and update to [3.10.6](https://pypi.org/project/awsebcli/3.10.6/) in separate commits, is that the convention here?

FRidh commented on pkgs/tools/virtualization/awsebcli/default.nix:
```diff
@@ -0,0 +1,89 @@
+  localPythonPackageSet = python.override {
+    packageOverrides = self: super: rec {
+      docker = super.docker.overridePythonAttrs (oldAttrs: rec {
+          pname = "docker-py";
```
you can use oldAttrs

eqyiel replied:
Unfortunately the name of the Pypi package changed in 2.0: https://github.com/docker/docker-py/releases/tag/2.0.0

`docker-py` is the name of the package prior to that but the `pname` from `oldAttrs` is `docker`.

Unless you're saying to use `${oldAttrs.pname}-py`?

FRidh commented on pkgs/tools/virtualization/awsebcli/default.nix:
```diff
@@ -0,0 +1,89 @@
+    packageOverrides = self: super: rec {
+      docker = super.docker.overridePythonAttrs (oldAttrs: rec {
+          pname = "docker-py";
+          version = "1.7.2";
```
note that you will have to update `name` as well yourself

eqyiel replied:
Thanks!

FRidh commented on pkgs/tools/virtualization/awsebcli/default.nix:
```diff
@@ -0,0 +1,89 @@
+    sha256 = "1g53z2flhp3navdf8lw6rgh99akf3k0ng1zkkqswvh66zswkxnwn";
+  };
+
+  buildInputs = with localPythonPackageSet.pkgs; [
```
these are likely inputs needed only for testing. The `buildPython*` functions supports `checkInputs`.

eqyiel replied:
Thanks, i wasn't aware that was a thing

FRidh replied:
Its only used in `buildPython*`, and I think only exists since 17.03.

FRidh reviewed (changes requested):
> I plan to add myself as a maintainer for the new packages and update to 3.10.6 in separate commits, is that the convention here?

Actually, the convention is one commit per expression you touch.

Commit:
maintainers.nix: Add myself

Commit:
pythonPackages.blessed: init at 1.14.2

This is a dependency of awsebcli.

Commit:
pythonPackages.cement: init at 2.8.2

This is a dependency of awsebcli.

Commit:
pythonPackages.tabulate: use fetchPypi

In order to make it nicer to override when a specific version is required for
Python derivations.

Commit:
pythonPackages.semantic-version: use fetchPypi

In order to make it nicer to override when a specific version is required for
Python derivations.

Commit:
awsebcli: init at 3.10.5

eqyiel commented:
> Actually, the convention is one commit per expression you touch.

I've split it out into more atomic commits, is this preferable? 😊

FRidh commented on pkgs/tools/virtualization/awsebcli/default.nix:
```diff
@@ -0,0 +1,99 @@
+        };
+      });
+
+      docker = super.docker.overridePythonAttrs (oldAttrs: rec {
```
am I correct that you are now using a different package for `docker`? That is, `docker-py` instead of `docker`?

eqyiel replied:
It's the same package from the same repo (https://github.com/docker/docker-py) but since version 2.0.0 (the version in pkgs/development/python-modules/docker.nix is 2.5.1) they changed the name that you would use to install it with pip.

It's in the release notes here: https://github.com/docker/docker-py/releases/tag/2.0.0

> The name of the pip package is now docker (was: docker-py). New versions of this library will only be published as docker from now on.

Should I create a new derivation for it?

FRidh commented on pkgs/tools/virtualization/awsebcli/default.nix:
```diff
@@ -0,0 +1,99 @@
+      colorama = super.colorama.overridePythonAttrs (oldAttrs: rec {
+        version = "0.3.7";
+
+        src = fetchPypi {
```
Note that `fetchPypi` is overridable, so you can write 
```
src = oldAttrs.src.override { inherit version; sha256 = "..."};
```
No need to change it, though.

FRidh commented on pkgs/tools/virtualization/awsebcli/default.nix:
```diff
@@ -0,0 +1,99 @@
+      docker = super.docker.overridePythonAttrs (oldAttrs: rec {
+          pname = "docker-py";
+          version = "1.7.2";
+          name = "${pname}-${version}";
```
This line is necessary for all the overrides since otherwise the old names will still be used.

Hopefully, after the release of 17.09, we can go further and get rid of `name` for `buildPython*` :)

eqyiel replied:
Thanks for letting me know!

FRidh reviewed (changes requested):
> I've split it out into more atomic commits, is this preferable?

Yep, much nicer this way!

eqyiel commented:
@FRidh was there something remaining to be done here?

eqyiel commented:
Rebased!

eqyiel commented:
Bump.  @FRidh is there something remaining to be done here?

eqyiel commented:
@FRidh @edolstra ping

FRidh commented:
I've pushed modified versions of these commits, e.g. b1c4498b415f3b9cfec8f619c749eccbb5314e66

FRidh commented:
My apologies for the long delay.

eqyiel commented:
@FRidh that's OK, thank you for your time!
